drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,1597 @@
+		info2_reg &= ~info2_poll_flag;
+		rcar_mmc_write_reg32(cfg, RCAR_MMC_INFO2, info2_reg);
+
+		for (w_off = 0; (w_off * sizeof(*buf)) < data->block_size; w_off++) {
+			if (is_read) {
+				buf[w_off] = sys_read64(cfg->reg_addr + RCAR_MMC_BUF0);
+			} else {
+				sys_write64(buf[w_off], cfg->reg_addr + RCAR_MMC_BUF0);
+			}
+		}
+	}
+#endif
+	ret = rcar_mmc_poll_reg_flags_check_err(cfg, RCAR_MMC_INFO1,
+					RCAR_MMC_INFO1_CMP, RCAR_MMC_INFO1_CMP, true);
+	if (ret) {
+		return ret;
+	}
+
+	/* clear access end flag  */
+	info1_reg = rcar_mmc_read_reg32(cfg, RCAR_MMC_INFO1);
+	info1_reg &= ~RCAR_MMC_INFO1_CMP;
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_INFO1, info1_reg);
+
+	return ret;
+}
+
+/**
+ * @brief Send command to MMC
+ *
+ * Sends a command to the MMC controller.
+ *
+ * @param dev MMC device
+ * @param cmd MMC command
+ * @param data MMC data. Leave NULL to send SD command without data.
+ *
+ * @retval 0 command was sent successfully
+ * @retval -ETIMEDOUT: command timed out while sending
+ * @retval -ENOTSUP: host controller does not support command
+ * @retval -EIO: I/O error
+ * @retval -EILSEQ: communication out of sync
+ */
+static int rcar_mmc_request(const struct device *dev,
+	struct sdhc_command *cmd,
+	struct sdhc_data *data)
+{
+	int ret;
+	const struct mmc_rcar_cfg *cfg;
+	uint32_t reg;
+	uint32_t response_type;
+	bool is_read = true;
+
+	if (!dev || !cmd || !dev->config) {
+		return -EINVAL;
+	}
+
+	/* TODO: add retries mechanism, use retries counter from cmd for that */
+
+	cfg = dev->config;
+	response_type = cmd->response_type & SDHC_NATIVE_RESPONSE_MASK;
+
+	if (rcar_mmc_card_busy(dev)) {
+		LOG_ERR("%s: device busy", dev->name);
+		return -EBUSY;
+	}
+
+	rcar_mmc_reset_and_mask_irqs(cfg);
+
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_INFO1_MASK, (uint32_t)~RCAR_MMC_INFO1_RSP);
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_INFO2_MASK, (uint32_t)~RCAR_MMC_INFO2_ERRORS);
+
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_ARG, cmd->arg);
+
+	reg = cmd->opcode;
+
+	if (data) {
+		rcar_mmc_write_reg32(cfg, RCAR_MMC_SIZE, data->block_size);
+		rcar_mmc_write_reg32(cfg, RCAR_MMC_SECCNT, data->blocks);
+		reg |= rcar_mmc_gen_data_cmd(cmd, data);
+		is_read = (reg & RCAR_MMC_CMD_RD) ? true : false;
+	}
+
+	ret = rcar_mmc_convert_sd_to_mmc_resp(response_type);
+	if (ret < 0) {
+		return -EINVAL;
+	}
+
+	reg |= ret;
+
+	LOG_DBG("(SD_CMD=%08x, SD_ARG=%08x)", cmd->opcode, cmd->arg);
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_CMD, reg);
+
+	/* wait until response end flag is set or errors occur */
+	ret = rcar_mmc_poll_reg_flags_check_err(cfg, RCAR_MMC_INFO1,
+				   RCAR_MMC_INFO1_RSP, RCAR_MMC_INFO1_RSP, true);
+	if (ret) {
+		return ret;
+	}
+
+	/* clear response end flag */
+	reg = rcar_mmc_read_reg32(cfg, RCAR_MMC_INFO1);
+	reg &= ~RCAR_MMC_INFO1_RSP;
+	rcar_mmc_write_reg32(cfg, RCAR_MMC_INFO1, reg);
+
+	rcar_mmc_extract_resp(cfg, cmd, response_type);
+
+	if (response_type != SD_RSP_TYPE_NONE) {
+		/* TODO: add cheching of RESP10 register here */
```

Comment:
Might as well add `LOG_WARN` to help user debug the issue

Reply:
user don't need to know about it
some of statuses are parsed by SDHC and they can produce some logs in case of errors

Reply:
Under "user" I meant driver user, not end-user. Basically, will  rest of the code work correctly if `response_type != SD_RSP_TYPE_NONE`?

Reply:
Better to change condition here to the next one:
```
if (response_type == SD_RSP_TYPE_R1 || response_type == SD_RSP_TYPE_R5 || response_type == SD_RSP_TYPE_R6 || response_type = SD_RSP_TYPE_R7 ||) {
```
because we need to check only `RCAR_MMC_CMD_RSP_R1` (we can get this bit directly from CMD reg and avoid this long line of conditions).

When some of error bits are set in Card status register (R1), it is meant that a CMD has been handled incorrectly or even isn't handled at all.

You can find an enumerator in Zephyr SD subsystem with all possible flags inside R1/Card status register, look at: `enum sd_r1_status`:
```
	/* Bits 0-2 reserved */
	SD_R1_AUTH_ERR = BIT(3),
	/* Bit 4 reserved for SDIO */
	SD_R1_APP_CMD = BIT(5),
	SD_R1_FX_EVENT = BIT(6),
	/* Bit 7 reserved */
	SD_R1_RDY_DATA = BIT(8),
	SD_R1_CUR_STATE = (0xFU << 9),
	SD_R1_ERASE_RESET = BIT(13),
	SD_R1_ECC_DISABLED = BIT(14),
	SD_R1_ERASE_SKIP = BIT(15),
	SD_R1_CSD_OVERWRITE = BIT(16),
	/* Bits 17-18 reserved */
	SD_R1_ERR = BIT(19),
	SD_R1_CC_ERR = BIT(20),
	SD_R1_ECC_FAIL = BIT(21),
	SD_R1_ILLEGAL_CMD = BIT(22),
	SD_R1_CRC_ERR = BIT(23),
	SD_R1_UNLOCK_FAIL = BIT(24),
	SD_R1_CARD_LOCKED = BIT(25),
	SD_R1_WP_VIOLATION = BIT(26),
	SD_R1_ERASE_PARAM = BIT(27),
	SD_R1_ERASE_SEQ_ERR = BIT(28),
	SD_R1_BLOCK_LEN_ERR = BIT(29),
	SD_R1_ADDR_ERR = BIT(30),
	SD_R1_OUT_OF_RANGE = BIT(31),
```

But, bare in mind, that not all of error bits can be affected by every SD CMD, there is a `Card status field/command—cross reference` table in `JEDEC Standard No. 84-A441`.

SDHC subsystem handles a lot of responses, but according to Usage examples of command from `R-Car S4 Series User’s Manual: Hardware` we should add checking of errors flags in place where I have put the comment.

Reply:
ok, I'll fix it

Reply:
upd: for some of commands we need to check rsp54

Reply:
My main point is that there is no sense of having `if` with empty body.  

It is right thing to check incoming data for consistency. No one argues with that. But, if a program sees that input data is invalid - it should do something.  Ideally - handle an error. If it can't  handle an error - it should propagate it to the next level of abstraction. At very least it should warn user and switch to a safe state. If you are certain that you encountered an error, you should do something.

Reply:
I've added a new commit which fixes it

Reply:
discussed, I've deleted the last commit,
we have agreed that it should be done inside SDHC subsystem, because checks are generic